### DISCUSSION TITLE
How to use multiple handlers for the same route depending on accept header?

### DISCUSSION THREAD

**Discussion by Grafcube** [tokio-rs/axum#3026] [Q&A] [answered]
### Summary

 Say I have the route `/user` with two handlers. I want to use `get_user_data` if the accept header is set to `application/json` and `user_page` otherwise. How do I do this?

My current set up uses a separate router with `ValidateRequestHeaderLayer` for multiple routes and those same routes are defined again outside that nested router. The problem is that it panics because the route has already been registered.

```rust
let apub_router = Router::new()
    .route("/user/:id", get(get_user))
    .route_layer(ValidateRequestHeaderLayer::accept(FEDERATION_CONTENT_TYPE));

let app = Router::new()
    .leptos_routes_with_context( // This generates /user/:id which conflicts with `apub_router`
        &uiconf.leptos_options,
        routes,
        {
            let state = instance_metadata.clone();
            move || provide_context(state.clone())
        },
        App,
    )
    .with_state(uiconf.leptos_options.clone())
    .nest("/", apub_router)
```

Any help is appreciated!

### axum version

0.7.7

**Comment by yanns** [comment]
A possibility is to handle this on the layer level: https://docs.rs/axum/latest/axum/response/index.html#returning-different-response-types

I don't know if you have many layers, and if it's a valid approach for you.

**Reply by Grafcube** [reply]
That doesn't work for me because I have multiple routes like this and I can't merge the handlers. The same route either returns html or json depending on the accept header. I can't use two different paths for this.

**Reply by yanns** [reply]
You could have something like:
```rust
async fn user_handler(...) -> Response {
  if accept_json(req) {
    user_json_handler(...)
  } else {
    user_html_handler(...)
  }
}

async fn user_html_handler(...) -> Response {
...
}

async fn user_json_handler(...) -> Response {
...
}
```
and have only the route `/user` using `user_handler`.
But I guess you don't like the boilerplate code.

**Reply by Grafcube** [reply]
I should have mentioned this in my original post, the html routes are being generated by leptos so I have no way to manually call it (that I know of). I only have the handler that needs to be behind the accept header.

**Reply by yanns** [reply]
Thanks for the clarification. I don't have more knowledge how you could get it.
Maybe someone else can help here.

**Reply by jplatte** [reply]
Can you update your question with some actual code snippets, indicating what is generated (unchangeable) then? And leave a comment on this thread once it's updated, as editing generally doesn't generate notifications.

**Reply by Grafcube** [reply]
I've edited my original post.

**Comment by mladedav** [comment, marked as answer]
This seems like another case similar to #1654 

What you can do now is nesting the router under a prefix and then use [middleware around the `Router`](https://docs.rs/axum/latest/axum/middleware/index.html#rewriting-request-uri-in-middleware) to rewrite the URL by prepending the same prefix based on your headers.

**Reply by Grafcube** [reply]
This worked for me, thanks!

```rust
let app = Router::new()
    .leptos_routes_with_context(
        &uiconf.leptos_options,
        routes,
        {
            let state = instance_metadata.clone();
            move || provide_context(state.clone())
        },
        App,
    )
    .with_state(uiconf.leptos_options.clone())
    .nest("/apub", apub_router)

let apub_middleware = axum::middleware::map_request(apub_middleware);
let app = apub_middleware.layer(app);
let app = ServiceExt::<Request>::into_make_service(app);
```

```rust
async fn apub_middleware<B>(mut request: Request<B>) -> Request<B> {
    if let Some(accept) = request.headers().get(ACCEPT) {
        if let Ok(accept) = accept.to_str() {
            if accept == FEDERATION_CONTENT_TYPE {
                let path = request.uri().path();
                if !path.starts_with("/apub") {
                    let new_path = Uri::builder()
                        .path_and_query(format!("/apub{}", path))
                        .build()
                        .unwrap();
                    *request.uri_mut() = new_path;
                }
            }
        }
    }
    request
}
```
